### PR TITLE
[State Sync] Add Execution Data ID Verification

### PR DESCRIPTION
**Pull request by peterargue**
Closes: https://github.com/onflow/flow-go/issues/2120

This is a refactor of https://github.com/onflow/flow-go/pull/2647, which was difficult to rebase due to its age.

This PR adds verification for the `ExecutionDataID` field in `ExecutionResults`.

### CODE REVIEW

File: module/chunks/chunkVerifier.go [hunk 1]
```diff
@@ -75,6 +78,14 @@ func (fcv *ChunkVerifier) Verify(
 		transactions = []*fvm.TransactionProcedure{
 			fvm.Transaction(txBody, vc.TransactionOffset+uint32(0)),
 		}
+
+		// enhance ChDP with collection if this is a system chunk. the collection is required to verify
+		// the chunk's execution data
+		if vc.ChunkDataPack.Collection == nil {
```
Comment (peterargue) [inline module/chunks/chunkVerifier.go:84]: I'm not sure if this is required in production. it was needed for tests.

File: ledger/complete/ledger.go [hunk 2]
```diff
@@ -200,7 +200,11 @@ func (l *Ledger) Set(update *ledger.Update) (newState ledger.State, trieUpdate *
 	// TODO: add test case
 	if update.Size() == 0 {
 		// return current state root unchanged
-		return update.State(), nil, nil
+		return update.State(), &ledger.TrieUpdate{
```
Comment (peterargue) [inline ledger/complete/ledger.go:203]: is this needed?

File: model/flow/chunk.go [hunk 3]
```diff
@@ -1,5 +1,9 @@
 package flow
 
+import (
+	"github.com/ipfs/go-cid"
```
Comment (ramtinms) [inline model/flow/chunk.go:4]: Just to confirm this is the package we were planning to use, I'm a bit worried that major entities on flow be dependent on the external libs, given that a simple change of order of fields or other facts would impact us heavily

Comment (peterargue) [reply model/flow/chunk.go:4]: This is the package used by bitswap for the data IDs. IPFS also uses it as the address for their data blocks. The core data structure is just a string and the serialization is versioned: https://github.com/ipfs/go-cid/blob/d900d766aff6fc5b91ba8d74efc99ae672701e84/cid.go#L170-L173

With that said, this is a fair point. We could convert these into `flow.Identifier` before adding it to the model, but I don't think would address the concern here. This PR does also add some tests which should catch basic changes:
https://github.com/onflow/flow-go/pull/4586/files#diff-1b92b673833800e45ffd8a743b0958aae4c05a9bfc0ad1780cb70a63c97e4c6dR124-R173

File: model/flow/chunk.go [hunk 4]
```diff
@@ -149,3 +156,15 @@ func (cl ChunkList) ByIndex(i uint64) (*Chunk, bool) {
 func (cl ChunkList) Len() int {
 	return len(cl)
 }
+
+// BlockExecutionDataRoot represents the root of a serialized execution_data.BlockExecutionData.
+// The hash of the serialized BlockExecutionDataRoot is the ExecutionDataID used within an
+// flow.ExecutionResult.
+type BlockExecutionDataRoot struct {
+	// BlockID is the ID of the block who's result this execution data is for.
```
Comment (AlexHentschel) [inline model/flow/chunk.go:164]:
```suggestion
	// BlockID is the ID of the block, whose result this execution data is for.
```

File: module/chunks/chunkVerifier.go [hunk 5]
```diff
@@ -306,5 +322,50 @@ func (fcv *ChunkVerifier) verifyTransactionsInContext(
...
+		TrieUpdate: trieUpdate,
+	}
+
+	cidProvider := provider.NewExecutionDataCIDProvider(execution_data.DefaultSerializer)
+	cedCID, err := cidProvider.CalculateChunkExecutionDataID(gocontext.Background(), chunkExecutionData)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to calculate CID of ChunkExecutionData: %w", err)
+	}
+
+	// 3. check that chunk execution data ID for our index is valid
```
Comment (AlexHentschel) [inline module/chunks/chunkVerifier.go:350]:
```suggestion
	// 3. check that with the chunk execution results that we created locally, 
	// we can reproduce the ChunkExecutionData's ID, which the execution node is stating in its ChunkDataPack
```

File: module/chunks/chunkVerifier.go [hunk 6]
```diff
@@ -306,5 +322,50 @@ func (fcv *ChunkVerifier) verifyTransactionsInContext(
...
+			execResID,
+		), nil
+	}
+
+	// 4. check the execution data root ID by calculating it using the provided execution data root
+	executionDataID, err := cidProvider.CalculateExecutionDataRootID(gocontext.Background(), chunkDataPack.ExecutionDataRoot)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to calculate ID of ExecutionDataRoot: %w", err)
+	}
+
```
Comment (AlexHentschel) [inline module/chunks/chunkVerifier.go:365]:
```suggestion
```

**Review by AlexHentschel — APPROVED** [review]
Thanks for cleaning this up, updating it to the latest version of the code and getting it over the finish line. Nice work. Most of my comments are of aesthetical nature ... hopefully not too much work. 

A couple additional suggestions:

- [x] [done] 
   can we add a brief documentation to the following field, please? https://github.com/onflow/flow-go/blob/7c784937638534e2ce6ed3d76fdbd51c93399d89/model/flow/execution_result.go#L17 Something like the following would already by fine
   ```golang
   ExecutionDataID  Identifier // hash commitment to `flow.BlockExecutionDataRoot`
   ``` 
- [x] [done 👉  https://github.com/onflow/flow-go/pull/4586/commits/e7a00d3a96688a82c44f58d4daada539a7660e65]
    Can we please extend the documentation of https://github.com/onflow/flow-go/blob/7c784937638534e2ce6ed3d76fdbd51c93399d89/model/flow/chunk.go#L163 I think it is worth-while to explain that 
   * The trie updates only contain the writes, which is helpful for clients to truslessly replicate the state. 
   * In comparison, the  chunk data packs contains all the register values at the chunk's starting state that were touched (written and/or read). This is necessary for Verification nodes to re-run chunk the computation.

File: model/chunks/executionDataFaults.go
[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go:1]: `Verify` differentiates `ChunkFaults` and errors. Errors indicate that verification failed for an internal reason, whereas `ChunkFaults` indicate it failed due to an issue with the chunk. When a `ChunkFault` is returned, no error is returned.

In the code that handles the response:
https://github.com/onflow/flow-go/blob/f1f4022999c8a4ef9bbf6d4d37de58aeee46db9b/engine/verification/verifier/engine.go#L191-L233

the fault message is used in a log message (and evetually a challenge), but no error is ultimate returned.

[the remark this answers is not in the document]
Comment (AlexHentschel) [reply model/chunks/executionDataFaults.go:1]: I think that is my point: we largely use [the convention](https://github.com/onflow/flow-go/blob/master/CodingConventions.md#error-handling) that byzantine inputs result in dedicated sentinel errors (which can carry auxiliary information). The code differentiates between unexpected (internal reasons) and expected failures (incl. byzantine inputs) via checking the error type. 

This code deviates from our convention, which I don't understand the reasons for. Just think about all our APIs, which generally can error due to internal problems and external malicious inputs. Splitting up all the errors would enormously bloat the interfaces in my opinion.

[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go:1]: bah, I changed my mind and deleted my comment before seeing your response. Updated in [56774a1](https://github.com/onflow/flow-go/pull/4586/commits/56774a1a719b79396e385823d767b988cf8730b1)

File: module/chunks/chunkVerifier.go [hunk 7]
```diff
@@ -75,6 +77,14 @@ func (fcv *ChunkVerifier) Verify(
 		transactions = []*fvm.TransactionProcedure{
 			fvm.Transaction(txBody, vc.TransactionOffset+uint32(0)),
 		}
+
+		// enhance ChDP with collection if this is a system chunk. the collection is required to verify
```
[the remark this answers is not in the document]
Comment (peterargue) [reply module/chunks/chunkVerifier.go]: Looking at the code, execution nodes do return `nil` for the collection in chunk data packs for the system chunk:
https://github.com/onflow/flow-go/blob/f1f4022999c8a4ef9bbf6d4d37de58aeee46db9b/engine/execution/block_result.go#L187-L206

So I'm guessing this was the get the tests working. I updated the implementation to verify the collection is not passed in [a240f20](https://github.com/onflow/flow-go/pull/4586/commits/a240f20e49df3b198d13ca88f30895b0266c4404)

**Review by yhassanzadeh13 — APPROVED** [review]

**Review by ramtinms — APPROVED** [review]

File: engine/execution/block_result.go [hunk 8]
```diff
@@ -12,6 +12,7 @@ type BlockExecutionResult struct {
 	*entity.ExecutableBlock
 
 	collectionExecutionResults []CollectionExecutionResult
+	ExecutionDataRoot          *flow.BlockExecutionDataRoot
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/execution/block_result.go]: addressed in dd1aa2f

File: engine/execution/computation/execution_verification_test.go [hunk 9]
```diff
@@ -851,9 +851,8 @@ func executeBlockAndVerifyWithParameters(t *testing.T,
 	require.Len(t, vcds, len(txs)+1) // +1 for system chunk
 
 	for _, vcd := range vcds {
-		_, fault, err := verifier.Verify(vcd)
+		_, err := verifier.Verify(vcd)
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/execution/computation/execution_verification_test.go]: addressed in dd1aa2f

File: engine/verification/verifier/engine.go [hunk 10]
```diff
@@ -188,47 +188,46 @@ func (e *Engine) verify(ctx context.Context, originID flow.Identifier,
...
-	spockSecret, chFault, err := e.chVerif.Verify(vc)
+	spockSecret, err := e.chVerif.Verify(vc)
 	span.End()
-	// Any err means that something went wrong when verify the chunk
-	// the outcome of the verification is captured inside the chFault and not the err
+
 	if err != nil {
-		return fmt.Errorf("cannot verify chunk: %w", err)
-	}
+		if !chmodels.IsChunkFaultError(err) {
```
[the remark this answers is not in the document]
Comment (peterargue) [reply engine/verification/verifier/engine.go:195]: addressed in dd1aa2f

File: model/chunks/executionDataFaults.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+// ExecutionResultID returns the execution result identifier including the faulty chunk
+func (cf CFExecutionDataInvalidChunkCID) ExecutionResultID() flow.Identifier {
+	return cf.execResID
+}
+
+// NewCFExecutionDataInvalidChunkCID creates a new instance of Chunk Fault (NewCFExecutionDataInvalidChunkCID)
+func NewCFExecutionDataInvalidChunkCID(
+	executionDataRootChunkCID cid.Cid,
+	computedChunkCIO cid.Cid,
```
[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go]: addressed in dd1aa2f

File: model/chunks/executionDataFaults.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+	chunkIndex                     uint64          // chunk's index
+	execResID                      flow.Identifier // chunk's ExecutionResult identifier
+	executionDataRootChunkLength   int             // number of ChunkExecutionDataIDs in ExecutionDataRoot
+	executionResultChunkListLength int             // number of chunks in ExecutionResult
+}
+
+var _ ChunkFaultError = (*CFExecutionDataChunksLengthMismatch)(nil)
+
+func (cf CFExecutionDataChunksLengthMismatch) String() string {
+	return fmt.Sprintf("execution data root chunk length (%d) is different then execution result chunk list length (%d) for chunk %d with result ID %s",
```
[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go]: addressed in dd1aa2f

File: model/chunks/executionDataFaults.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+	chunkIndex                uint64          // chunk's index
+	execResID                 flow.Identifier // chunk's ExecutionResult identifier
+	executionDataRootChunkCID cid.Cid         // ExecutionDataRoot's CID for the chunk
+	computedChunkCID          cid.Cid         // computed CID for the chunk
+}
+
+var _ ChunkFaultError = (*CFExecutionDataInvalidChunkCID)(nil)
+
+func (cf CFExecutionDataInvalidChunkCID) String() string {
+	return fmt.Sprintf("execution data chunk CID (%s) is different then computed (%s) for chunk %d with result ID %s",
```
[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go]: addressed in dd1aa2f

File: model/chunks/executionDataFaults.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
+package chunks
```
[the remark this answers is not in the document]
Comment (peterargue) [reply model/chunks/executionDataFaults.go:1]: addressed in dd1aa2f

File: module/chunks/chunkVerifier.go [hunk 15]
```diff
@@ -283,28 +300,82 @@ func (fcv *ChunkVerifier) verifyTransactionsInContext(
...
+	// 2. build our chunk's chunk execution data using the locally calculated values, and calculate
+	// its CID
+	chunkExecutionData := execution_data.ChunkExecutionData{
+		Collection: cedCollection,
+		Events:     events,
+		TrieUpdate: trieUpdate,
+	}
+
+	cidProvider := provider.NewExecutionDataCIDProvider(execution_data.DefaultSerializer)
+	cedCID, err := cidProvider.CalculateChunkExecutionDataID(gocontext.Background(), chunkExecutionData)
```
[the remark this answers is not in the document]
Comment (peterargue) [reply module/chunks/chunkVerifier.go]: addressed in dd1aa2f

**Pull request merged**